I have added some support for the Sync function without talking to mount.

damianofasoli opened the pull request:
I saw (serial sniffer), at least with my Exos II mount, that ASCOM driver does not send commands to mount with SYNC function.

So, I managed the idea to change reported coordinates from driver comparing the position of the last GOTO (called Sync base coordinates) with that of the SYNC function (later), and moving coordinates accordingly.

Later I extended this behaviour to motion and than to tracking (mount tracks too fast or too lost) removing the one from motion (covered by the one in tracking)

KStars seems to behave well.

I'll try with a real night observation session, sooner
And also to do something with the handbox to see how it interacts.
For now the idea was to use only KStars
1 Goto
2 Some motions to align telescop, or even StellarSolver
3 Sync

If you believe that these integrations could be useful for your project, I'll feel glad to have contributed

Nice day
D. Fasoli

damianofasoli committed:
.

damianofasoli committed:
Sync integration without talking to mount

damianofasoli committed:
.

damianofasoli committed:
.

damianofasoli committed:
.

damianofasoli committed:
.

damianofasoli committed:
.

damianofasoli committed:
.

damianofasoli committed:
Align Sync Base when MotionToDirection ends

damianofasoli committed:
.

damianofasoli committed:
Align Sync Base while tracking: motions occurred, mount tracking not perfect

damianofasoli committed:
.

damianofasoli committed:
.

damianofasoli committed:
.

damianofasoli committed:
.

kneo commented on ExosIIMountControl.hpp:
```diff
@@ -415,11 +444,33 @@ class ExosIIMountControl :
-                           &messageBuffer[0], 0, messageBuffer.size());
-                //return rc && mMountStateMachine.DoTransition(TelescopeSignals::GoTo);
+            {                
+                if (true) {  
```
If you are sure this code is fine, just remove the old, I don't see a reason for this `if(true)` here.

kneo commented:
I'm glad you contributed, since I'm cursed with bad weather, I can not test this myself, therefore, I'd like to keep this PR open until you are sure for its improvement.

If you are sure the code works, I'm going to accept the PR, and push the changes into the indi-3rdparty repository.

damianofasoli commented:
Hello

Thank for your interest
I'll try on a live night the function really
And then I'll let you know

Regards
DF


On Sat, Jan 27, 2024, 19:31 Kevin Krüger ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In CMakeLists.txt
> <https://github.com/kneo/indi-bresserexos2/pull/12#discussion_r1468547521>
> :
>
> > @@ -1,5 +1,5 @@
>  cmake_minimum_required(VERSION 3.0)
> -project(indi-bresserexos2 VERSION 0.901)
> +project(indi-bresserexos2 VERSION 9.999)
>
> A smaller version number would be appreciated, 0.950 or something.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kneo/indi-bresserexos2/pull/12#pullrequestreview-1847152219>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFUOZWXGASUCIA5YMSJYRZ3YQVBY5AVCNFSM6AAAAABCNNUVUWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNBXGE2TEMRRHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

damianofasoli commented:
I have put.a fake version only tio have a difference in Indi Control Panel

The real version will be of your choice of course

Regards
DF



On Sat, Jan 27, 2024, 19:31 Kevin Krüger ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In CMakeLists.txt
> <https://github.com/kneo/indi-bresserexos2/pull/12#discussion_r1468547521>
> :
>
> > @@ -1,5 +1,5 @@
>  cmake_minimum_required(VERSION 3.0)
> -project(indi-bresserexos2 VERSION 0.901)
> +project(indi-bresserexos2 VERSION 9.999)
>
> A smaller version number would be appreciated, 0.950 or something.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kneo/indi-bresserexos2/pull/12#pullrequestreview-1847152219>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFUOZWXGASUCIA5YMSJYRZ3YQVBY5AVCNFSM6AAAAABCNNUVUWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNBXGE2TEMRRHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

damianofasoli committed:
Some optimizations

damianofasoli commented:
I've done some optimizations and changed version as proposed

Still waiting for real night observations to try all

Regards
DF

damianofasoli commented:
Yesterday I tried the modifications with a live session with Kstars

Sync works, and I have halse plate solved successfull (Resolve, Sync and Slew) with StellarSolver Align functions a couple of time.

I believe that Sync in this way works using only skycharts applications commanding the mount 

For now, I have not tried at all combining motions with the handpad (firmware 2.3)
and tbe difference with 0.901 version

I believe however that only the telescope moves, without reporting motions to skychart applications
And so a sync to indi is required to the real position
There should be no problem in this
I'll try

damianofasoli commented:
Tried with the handpad
It moves the telescope using motion keys and coordinates change in skychart applications, reported by the Indi Driver
Same for GoTo
This is good

However if I try a Goto with the handpad it uses J2000 coordinates and so the position in skychart applications is different
In this case we need a Sync command to the driver
Perhaps doing a Sync from the Handpad it introduces the correction there and GoTo from the handpand will use ccordinates corrected by the Sync instead of J2000
This point is open

Therefore no problems with handpads too

The overall problem of my Sync function to Indi driver is that it creates different from coordinates by driver and the handpad. But it happens often, I read, and this is not a problem if the handpand is not used at all

Regards
DF

kneo commented:
Not being forced to use the hand pad would be a great improvement indeed. 

There is an reverse engineered firmware for the handpad, you may check it out : 
[https://github.com/Spitzbube/EXOS-2_GoTo_HandController](https://github.com/Spitzbube/EXOS-2_GoTo_HandController)

May be you can fix problems there as well.

damianofasoli committed:
In x86 systems (where GCC defaults fo -fsigned-char) there was a problem reading some bytes from mount (casted to negative values).
No problem fith ARM systems (where GCC defaults to -funsigned-char)

damianofasoli commented:
Hi, sorry for late answer
The hand controller is out of my possibilities :-)

Proabably I'll compare the mount using ASCOM and Indi_BresserExos2 driver more closely.

I tried to use indi_bresserexos2 with a x86/x64 instead of arm and I discovered  that the input from mount was broken

Then (after hours of tries :-)  ) I discovered that the problem was the conversion from char to uint16_t with negative values on systems (GCC compilers) where char are signed 

You can see the whole matter here
https://gcc.gnu.org/onlinedocs/gcc-4.8.0/gcc/C-Dialect-Options.html
searching for
-funsigned-char

I have done a new commit on my branch, version 951

Regards
DF

damianofasoli committed:
.

damianofasoli committed:
Update CMakeLists.txt

Aligned with knro commits:
Use minimum CMake 3.16 (Dec 26, 2023)
Update to compile with INDI guide interface refactor (#961) (Aug 19, 2024)

damianofasoli commented:
Hello,

I saw that there were some updates in indi-3rdparty repository for this mount indi-bresserexos2

I have aligned it in my github branch

All seems OK using Indi 2.0.9 latest, both on Linux x86-64 and Ubuntu Arm aarch64

If you agree, it is possible to merge back my github branch on indi-3rdparty repository
The version will probably be 901 or something similar. 951 and 952 are just my tries to avoid confusion on components installed.

I realize that with my approach sync is simulated by the driver and not read back from mount.
It works however and it seems that also ASCOM driver does something similar.

However, at least, the fix introduced here below is important for x86-x64 architectures.
https://github.com/kneo/indi-bresserexos2/pull/12#issuecomment-1974964678

Let's stay in touch.

Nice day
D. Fasoli

pweingardt commented:
This may be not the right place to say it, but the changes made in damianofaolis fork made the driver work for me in the first place. Without those changes (default indi-bresserexos2 repository) the driver for my EXOS 2 mount does not work *at all*. The only thing not working right now is guiding through the mount itself (doesn't work with the ST4 cable either, so maybe user error on my side).
Thank you very much for your hard work.

damianofasoli commented:
Thanks you @pweingardt 
Di you refer to x86 fix or to Sync function or both?

Sync function Is emulated in the driver without sending commands to he telescope and for this reason I don't know if this fork can be committed to official repository.

However I have no more Exos2 mount 
Thanks again

damianofasoli commented:
@kneo, do you think it is possible to publish this merge request to indi-3rdparty repository?

Sync, we know, is only simulated inside the driver and not sent to telescope.
But with 2.3 Handbox firmware sync sent to telescope does not work at all; furthermore ASCOM driver seems to do something similar (no commands to handbox for sync) and in the documentation there is already the warning that coordinates of the driver may be different with coordinates of handbox (it just happen so with my sync review)

And a part from Sync, there is the big problem with x86/x64 versions, rather than ARM versions.
There the driver does not work at all 
This is an important fix for official version

Regards

damianofasoli commented:
There is another problem to fix, by the way.
The diagonal keys to move the mount both west and south for example, just as those on KStarts/Ekos for example, crash the driver!
Perhaps it is caused by changes in Indi implementations (start motion, stop motion) or something similar.

pweingardt commented:
The default indi_bresserexos2 driver does not work for me, it doesn't get the current orientation (RA, DEC) and it does not slew at all. I don't use those buttons, so it is not a problem for me.

damianofasoli commented:
Ok @pweingardt
The default driver works only for ARM systems (RaspBerry PI, ecc...)

For x86/x64 systems, my fix is needed.
You can see above the reason

@kneo, I believe the "pull request" in our forks can be committed to
default indi 3rdparty repository, if you agree.
This fix for x86/x64 versions is quite important!
And there is also sync, even though is not handled by the telescope but
simulated into the driver. It seems that also ASCOM behaves in this say,
and furthermore we know that coordinates may already be different between
indi driver and telescope handbox.

Regards

Il giorno dom 30 mar 2025 alle ore 21:54 pweingardt <
***@***.***> ha scritto:

> The default indi_bresserexos2 driver does not work for me, it doesn't get
> the current orientation (RA, DEC) and it does not slew at all. I don't use
> those buttons, so it is not a problem for me.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kneo/indi-bresserexos2/pull/12#issuecomment-2764719286>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFUOZWTMJ6UXBQWRQUGRACD2XBDZBAVCNFSM6AAAAAB2BTOT6SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRUG4YTSMRYGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: pweingardt]*pweingardt* left a comment (kneo/indi-bresserexos2#12)
> <https://github.com/kneo/indi-bresserexos2/pull/12#issuecomment-2764719286>
>
> The default indi_bresserexos2 driver does not work for me, it doesn't get
> the current orientation (RA, DEC) and it does not slew at all. I don't use
> those buttons, so it is not a problem for me.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kneo/indi-bresserexos2/pull/12#issuecomment-2764719286>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFUOZWTMJ6UXBQWRQUGRACD2XBDZBAVCNFSM6AAAAAB2BTOT6SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRUG4YTSMRYGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

kneo commented:
> Ok @pweingardt The default driver works only for ARM systems (RaspBerry PI, ecc...) For x86/x64 systems, my fix is needed. You can see above the reason @kneo, I believe the "pull request" in our forks can be committed to default indi 3rdparty repository, if you agree. This fix for x86/x64 versions is quite important! And there is also sync, even though is not handled by the telescope but simulated into the driver. It seems that also ASCOM behaves in this say, and furthermore we know that coordinates may already be different between indi driver and telescope handbox. Regards Il giorno dom 30 mar 2025 alle ore 21:54 pweingardt < ***@***.***> ha scritto:
> […](#)
> The default indi_bresserexos2 driver does not work for me, it doesn't get the current orientation (RA, DEC) and it does not slew at all. I don't use those buttons, so it is not a problem for me. — Reply to this email directly, view it on GitHub <[#12 (comment)](https://github.com/kneo/indi-bresserexos2/pull/12#issuecomment-2764719286)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFUOZWTMJ6UXBQWRQUGRACD2XBDZBAVCNFSM6AAAAAB2BTOT6SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRUG4YTSMRYGY> . You are receiving this because you authored the thread.Message ID: ***@***.***> [image: pweingardt]*pweingardt* left a comment ([kneo/indi-bresserexos2#12](https://github.com/kneo/indi-bresserexos2/pull/12)) <[#12 (comment)](https://github.com/kneo/indi-bresserexos2/pull/12#issuecomment-2764719286)> The default indi_bresserexos2 driver does not work for me, it doesn't get the current orientation (RA, DEC) and it does not slew at all. I don't use those buttons, so it is not a problem for me. — Reply to this email directly, view it on GitHub <[#12 (comment)](https://github.com/kneo/indi-bresserexos2/pull/12#issuecomment-2764719286)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFUOZWTMJ6UXBQWRQUGRACD2XBDZBAVCNFSM6AAAAAB2BTOT6SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONRUG4YTSMRYGY> . You are receiving this because you authored the thread.Message ID: ***@***.***>

Just go ahead. The driver is open source for a reason. Fork the third party repo, and merge it over. I no longer have this mount, and I feel uncomfortable committing code I didn't test personally to a repository that is commercially used.

damianofasoli commented:
Ok @Kneo, I'll try. Thanks for your trust

The updates here described have been tested with my mount with Handpad Firmware 2.3 and I'm sure they're working
Perhaps the original sync command was developed focusing previous firmware releases

I have no more Exos 2 mount too, and so following issues will be difficult to be developed and tested.

@pweingardt , have you got too the Firmware 2.3 on the Handpad?
If so, could you test the "sync" function once again? (plate solving with sync and move to target for example, or even simple sync standalone)
Thanks a lot!

pweingardt commented:
I have the firmware 2.3 on the handbox (it was preinstalled). Currently my workflow is as follows:

- Setup mount and camera (DSLR)
- Polar alignment with mount and camera and *manual slew* (automatic slew doesn't work)
- slew mount just slightly via Kstars
- Plate solve and sync (works only if I slew the mount to any target in the previuos step)
- slew to bright star (sync and slew to target) and setup focus with a bahtinov mask
- setup EKOS scheduler for a specific target and let it run

There are some things, that do not work:

- Sync mount without slewing at all
- set up different slew speeds for manual control (in EKOS, the handbox works just fine)
- meridian flip
- sometimes with the mosaic planner the mount slews and tries to take an image at the same time. It usually detects the slewing and aborts the plate solving. It is not that bad...
- synchronize the current position (RA, DEC) to the handbox

If the sky is clear tonight, I will try to set up guiding. So far I could not test this feature. Last time it didn't work at all, but maybe it was my fault.

damianofasoli commented:
Hi @pweingardt 

Ok, I'm aligned with your reports
Sync works only when the mount is in tracking mode, and so you have to slew the mount or perform a GoTo action before.
This is known, and you can find it in the documentation provided by @kneo
Before my update to the driver, sync didn't work at all from a planetarium software, neither in tracking mode. It only worked after Handpad movements, probably respect to the Handpad it self.
My update for sync function works only at driver level, simulating new coordinates. Handpad does not know them and continues to report previous coordinates. 
This is one of the main reasons we haven't already committed the update to the main 3rdparty repository, but as agreed with @kneo, I'll try to submit the update in the following days, since it is a really plus for the users who don't use at all the handpad and use only planetarium software.
Since you have already done "plate solving", when the mount is in tracking mode (known limit), I consider your test OK

As regards the other things which don't work
a) sync, we have described
b) the Indi driver supports only one speed mode
c) meridian flip is not supported by Indi driver, neither detection of west and east pier
d) this is a problem with Ekos, rather than the mount and the driver, I believe. I have the same problems with other mounts (EQ3 with OnStep Indi driver for example) and I solved it setting a larger "Saddle time" in Ekos.
It starts with 3000 milliseconds but you can set it to 5000 milliseconds for example, in the Plate Solving or Polar Alignment mask. In this way, the next image will be done later, when the mount has just finished slewing and vibrating
e) the Indi driver uses often different coordinates from the Handbox, as reported by @kneo. Using sync from planetarium software, after my update, this behaviour will be more present

I don't know if we'll be able to solve them
By the way, I have not the mount too, like @kneo, and tests will be difficult.
This a reason 

However the two updates I have made, will probably be merged into the main branch of the driver
a) indi driver usage with x86/x64 architectures and not only ARM ones
b) sync function from planetarium software (when the mount is tracking)

Regards
 
d)

damianofasoli commented:
Hi @pweingardt 

Ok, I'm aligned with your reports
Sync works only when the mount is in tracking mode, and so you have to slew the mount or perform a GoTo action before.
This is known, and you can find it in the documentation provided by @kneo
Before my update to the driver, sync didn't work at all from a planetarium software, neither in tracking mode. It only worked after Handpad movements, probably respect to the Handpad it self.
My update for sync function works only at driver level, simulating new coordinates. Handpad does not know them and continues to report previous coordinates. 
This is one of the main reasons we haven't already committed the update to the main 3rdparty repository, but as agreed with @kneo, I'll try to submit the update in the following days, since it is a really plus for the users who don't use at all the handpad and use only planetarium software.
Since you have already done "plate solving", when the mount is in tracking mode (known limit), I consider your test OK

As regards the other things which don't work
a) sync, we have described
b) the Indi driver supports only one speed mode
c) meridian flip is not supported by Indi driver, neither detection of west and east pier
d) this is a problem with Ekos, rather than the mount and the driver, I believe. I have the same problems with other mounts (EQ3 with OnStep Indi driver for example) and I solved it setting a larger "Saddle time" in Ekos.
It starts with 3000 milliseconds but you can set it to 5000 milliseconds for example, in the Plate Solving or Polar Alignment mask. In this way, the next image will be done later, when the mount has just finished slewing and vibrating
e) the Indi driver uses often different coordinates from the Handbox, as reported by @kneo. Using sync from planetarium software, after my update, this behaviour will be more present

I don't know if we'll be able to solve them
By the way, I have not the mount too, like @kneo, and tests will be difficult.
This a reason 

However the two updates I have made, will probably be merged into the main branch of the driver
a) indi driver usage with x86/x64 architectures and not only ARM ones
b) sync function from planetarium software (when the mount is tracking)

Regards
 
d)

damianofasoli commented:
Hello @kneo  and @pweingardt 

I have submitted the pull request to 3rdparty main repository just yesterday
Today Jasem has merged the pull request

So I believe that main repository is now aligned with our variations and we can compile drivers from sources there

Stay tuned
Regards

pweingardt commented:
Yesterday I was able to test the guiding functionality, and it worked. The RMS was okay for my telescope. 
Thank you for all your hard work!

damianofasoli commented:
Wonderful 

At this point, if you can try the official version once again, It Is superlative

But yoo'll have to compile It from sources, as uou have done with my fork

Thanks

pweingardt commented:
I just set up my new build and compiled kstars with all dependencies. The Besser EXOS2 driver worked right from the start, version 0.953, if I remember correctly. Thank you very much!
